lib/spack/spack/test/installer.py

```diff
@@ -252,6 +252,50 @@ def test_installer_str(install_mockery):
     assert "failed (0)" in istr
 
 
+def test_installer_prune_built_build_deps(install_mockery, monkeypatch, tmpdir):
+    r"""
+    Ensure that build dependencies of installed deps are pruned
+    from installer package queues.
+
+                (a)
+                / \
+               /   \
+             (d)   (b) <-- is installed already so we should
+               \   / \     prune (c) from this install since
+                \ /  (c)   it is *only* needed to build (b)
+                (e)
+
+    Thus since (b) is already installed our build_pq dag should
+    only include four packages. [(a), (b), (d), (e)]
+    """
+
+    @property
+    def _mock_installed(self):
+        return self.name in ["b"]
+
+    # Mock the installed property to say that (b) is installed
+    monkeypatch.setattr(spack.spec.Spec, "installed", _mock_installed)
+
+    # Create mock repository with packages (a), (b), (c), (d), and (e)
+    builder = spack.repo.MockRepositoryBuilder(tmpdir.mkdir("mock-repo"))
+
+    builder.add_package("a", dependencies=[("b", "build", None), ("d", "build", None)])
+    builder.add_package("b", dependencies=[("c", "build", None), ("e", "build", None)])
```

Comment:
Can you add another dependency `b -> f` with type `("build", "link")` to confirm that an edge is not pruned if it is needed via a non-build deptype (even if it is also a build dep)?

Reply:
`MockRepositoryBuilder` doesn't allow for a tuple of dependency types but I've added a package with the `"all"` dependency type as a substitute for `("build", "run", "link")`.

Reply:
That's probably something we should change on the `MockRepositoryBuilder`, but this should be fine for now